logic/atrium-logic-common/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultIterableLikeAssertions.kt

```diff
@@ -176,10 +174,47 @@ class DefaultIterableLikeAssertions : IterableLikeAssertions {
     ): Assertion = LazyThreadUnsafeAssertionGroup {
         val list = transformToList(container, converter)
 
-        val lookupHashSet = HashSet<E>()
-        val duplicates = createIndexAssertions(list) { (_, element) ->
-            !lookupHashSet.add(element)
-        }
+        val lookupHashMap = HashMap<E, Int>()
+        val duplicateIndices = HashMap<Int, Pair<E, MutableList<Int>>>()
+
+        list.asSequence()
+            .mapWithIndex()
+            .forEach { (index, element) ->
+                lookupHashMap[element]?.let {
+                    duplicateIndices.getOrPut(it) {
+                        Pair(element, mutableListOf<Int>())
+                    }.second.add(index)
+                } ?: let {
+                    lookupHashMap[element] = index
+                }
+            }
+
+        val duplicates = duplicateIndices.asSequence()
+            .map { (index, pair) ->
+                pair.let { (element, duplicateIndices) ->
+                    assertionBuilder.descriptive
+                        .failing
+                        .withFailureHint {
+                            assertionBuilder.explanatoryGroup
+                                .withDefaultType
+                                .withExplanatoryAssertion(
+                                    TranslatableWithArgs(
+                                        DescriptionIterableAssertion.DUPLICATED_BY,
+                                        duplicateIndices.joinToString(", ")
+                                    )
+                                )
+                                .build()
+                        }
+                        .showForAnyFailure
+                        .withDescriptionAndRepresentation(
+                            TranslatableWithArgs(DescriptionIterableAssertion.INDEX, index),
+                            element
+                        )
+                        .build()
+                }
+            }
+            .toList()
```

Comment:
I might be wrong but here `asSequence` does not pay off as you only have one transformation, to the contrary it is an unnecessary step IMO and can be simplified to
```suggestion
        val duplicates = duplicateIndices
            .map { (index, pair) ->
                pair.let { (element, duplicateIndices) ->
                    assertionBuilder.descriptive
                        .failing
                        .withFailureHint {
                            assertionBuilder.explanatoryGroup
                                .withDefaultType
                                .withExplanatoryAssertion(
                                    TranslatableWithArgs(
                                        DescriptionIterableAssertion.DUPLICATED_BY,
                                        duplicateIndices.joinToString(", ")
                                    )
                                )
                                .build()
                        }
                        .showForAnyFailure
                        .withDescriptionAndRepresentation(
                            TranslatableWithArgs(DescriptionIterableAssertion.INDEX, index),
                            element
                        )
                        .build()
                }
            }
```